### PR TITLE
BUG: HDFStore select for large integer

### LINKED ISSUE #53323: BUG: HDFStore select where condition doesn't work for large integer
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
s=pd.HDFStore('/tmp/test.h5')
df=pd.DataFrame(zip(['a','b','c','d'],[-9223372036854775801, -9223372036854775802,-9223372036854775803,123]), columns=['x','y'])
s.append('data',df,data_columns=True, index=False)
s.close()
s=pd.HDFStore('/tmp/test.h5')
s.select('data', where='y==-9223372036854775801')
```

### Issue Description

The above code will return no records which is obviously wrong. The issue is caused by the below in pandas/core/computation/pytables.py (function convert_value of class BinOp). Precision is lost after converting large integer to float and then back to integer:

        elif kind == "integer":
            v = int(float(v))
            return TermValue(v, v, kind)

### Expected Behavior

return the matched row instead of empty dataframe

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 37ea63d540fd27274cad6585082c91b1283f963d
python           : 3.10.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.15.0-72-generic
Version          : #79-Ubuntu SMP Wed Apr 19 08:22:18 UTC 2023
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.0.1
numpy            : 1.23.5
pytz             : 2022.1
dateutil         : 2.8.2
setuptools       : 59.6.0
pip              : 22.0.2
Cython           : 0.29.33
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.8.0
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 7.31.1
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : None
brotli           : 1.0.9
fastparquet      : None
fsspec           : 2023.1.0
gcsfs            : None
matplotlib       : 3.5.1
numba            : 0.56.4
numexpr          : 2.8.4
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 12.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.8.0
snappy           : None
sqlalchemy       : None
tables           : 3.8.0
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None

</details>

### PR DESCRIPTION
**Pull request by SanjithChockan**
- [x] closes #53323
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Used decimal conversion instead of float to prevent precision loss

### CODE REVIEW

File: doc/source/whatsnew/v2.1.0.rst [hunk 1]
```diff
@@ -478,10 +478,10 @@ Conversion
 - Bug in :func:`DataFrame.style.to_latex` and :func:`DataFrame.style.to_html` if the DataFrame contains integers with more digits than can be represented by floating point double precision (:issue:`52272`)
 - Bug in :func:`array`  when given a ``datetime64`` or ``timedelta64`` dtype with unit of "s", "us", or "ms" returning :class:`PandasArray` instead of :class:`DatetimeArray` or :class:`TimedeltaArray` (:issue:`52859`)
 - Bug in :meth:`ArrowDtype.numpy_dtype` returning nanosecond units for non-nanosecond ``pyarrow.timestamp`` and ``pyarrow.duration`` types (:issue:`51800`)
+- Bug in :meth:`BinOp.convert_value` loses precision when converting int to float if integer in DataFrame stored in HDFS file format is large (:issue:`54186`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:481]: Could you rewrite this in terms of a public pandas API (i.e. BInOp is not public)

File: pandas/tests/io/pytables/test_select.py [hunk 2]
```diff
@@ -943,3 +943,20 @@ def test_select_empty_where(tmp_path, where):
         store.put("df", df, "t")
         result = read_hdf(store, "df", where=where)
         tm.assert_frame_equal(result, df)
+
+
+def test_select_large_integer(tmp_path):
+    path = tmp_path / "large_int.h5"
+    s = HDFStore(path)
```
Comment (mroeschke) [inline pandas/tests/io/pytables/test_select.py:950]: Could you use the context manger version of this?

File: pandas/tests/io/pytables/test_select.py [hunk 3]
```diff
@@ -943,3 +943,20 @@ def test_select_empty_where(tmp_path, where):
...
+    df = DataFrame(
+        zip(
+            ["a", "b", "c", "d"],
+            [-9223372036854775801, -9223372036854775802, -9223372036854775803, 123],
+        ),
+        columns=["x", "y"],
+    )
+    s.append("data", df, data_columns=True, index=False)
+    assert (
+        s.select("data", where="y==-9223372036854775801").get("y").get(0)
```
Comment (mroeschke) [inline pandas/tests/io/pytables/test_select.py:960]: Could you assign `result` and `expected` variables to make this easier to read?

File: pandas/core/computation/pytables.py [hunk 4]
```diff
@@ -233,7 +233,16 @@ def stringify(value):
...
-            v = int(float(v))
+            from decimal import (
+                Decimal,
+                InvalidOperation,
+            )
+
+            try:
+                v_dec = Decimal(v)
+                v = int(v_dec.to_integral_exact(rounding="ROUND_HALF_EVEN"))
+            except InvalidOperation:
```
[the remark this answers is not in the document]
Comment (SanjithChockan) [reply pandas/core/computation/pytables.py:244]: One of the unit test passes in a string and only matches the ValueError raised when tried converting by float type. Initially I just had a `float(v)` to catch the error but didn't seem like good practice.

File: doc/source/whatsnew/v2.1.0.rst (hunk 1 above)
Comment (SanjithChockan) [reply doc/source/whatsnew/v2.1.0.rst]: Rewrote it referencing HDFStore select method.

File: pandas/core/computation/pytables.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/core/computation/pytables.py:244]: > One of the unit test passes in a string and only matches the ValueError raised when tried converting by float type.

Which unit test?

[the remark this answers is not in the document]
Comment (SanjithChockan) [reply pandas/core/computation/pytables.py:244]: pandas/tests/io/pytables/test_select.py::test_query_compare_column_type

[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/core/computation/pytables.py:244]: Okay. Could you rewrite this as

```
try:
   v_dec = Decimal(v)
except InvalidOperation:
    # raise original exception message
    float(v)
else:
   v = int(v_dec.to_integral_exact(rounding="ROUND_HALF_EVEN"))
```

File: pandas/core/computation/pytables.py [hunk 5]
```diff
@@ -233,7 +233,16 @@ def stringify(value):
                 result = metadata.searchsorted(v, side="left")
             return TermValue(result, result, "integer")
         elif kind == "integer":
-            v = int(float(v))
+            from decimal import (
```
Comment (mroeschke) [inline pandas/core/computation/pytables.py:236]: Can you move these imports to the top

File: pandas/core/computation/pytables.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SanjithChockan) [reply pandas/core/computation/pytables.py:244]: the `float(v)` inside `Decimal` would lose precision

```
>>> from decimal import Decimal
>>> v = -9223372036854775801
>>> v = Decimal(float(v))
>>> v
Decimal('-9223372036854775808')
>>> int(v)
-9223372036854775808
```

[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/core/computation/pytables.py:244]: Okay could you do

```
try:
    v_dec = Decimal(v)
except InvalidOperation:
    # raise original non-convertible float error
    float(v)
else:
    v = int(v_dec.to_integral_exact(rounding="ROUND_HALF_EVEN"))
```

[the remark this answers is not in the document]
Comment (SanjithChockan) [reply pandas/core/computation/pytables.py:242]: this works!

File: pandas/core/computation/pytables.py [hunk 6]
```diff
@@ -233,7 +237,12 @@ def stringify(value):
                 result = metadata.searchsorted(v, side="left")
             return TermValue(result, result, "integer")
         elif kind == "integer":
-            v = int(float(v))
+            try:
+                v_dec = Decimal(v)
+            except InvalidOperation:
```
Comment (mroeschke) [inline pandas/core/computation/pytables.py:242]: Could you add a comment where why `float(v)` is called?

Comment (SanjithChockan) [reply pandas/core/computation/pytables.py:242]: added comments for reason and referencing to PR.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @SanjithChockan

**Pull request merged**